activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1468,6 +1471,12 @@ def uniq!(name)
     # is passed in) are not instances of the same model that the relation is
     # scoping.
     def excluding(*records)
+      records.map! do |record|
+        case record
+        when Relation then record.select(record.primary_key)
+        else record
+        end
+      end
```

Comment:
(Continuing from https://github.com/rails/rails/pull/49570#discussion_r1354892418)

`select` will trigger a query even if the `Relation` is already `loaded?`.  It also allocates Active Record objects when we only really want the IDs.  I think we could use `ids` here instead if we first `extract!` the `Relation` elements:

```diff
--- a/activerecord/lib/active_record/relation/query_methods.rb
+++ b/activerecord/lib/active_record/relation/query_methods.rb
@@ -3,6 +3,7 @@
 require "active_record/relation/from_clause"
 require "active_record/relation/query_attribute"
 require "active_record/relation/where_clause"
+require "active_support/core_ext/array/extract"
 require "active_support/core_ext/array/wrap"
 
 module ActiveRecord
@@ -1471,20 +1472,16 @@ def uniq!(name)
     # is passed in) are not instances of the same model that the relation is
     # scoping.
     def excluding(*records)
-      records.map! do |record|
-        case record
-        when Relation then record.select(record.primary_key)
-        else record
-        end
-      end
+      relations = records.extract! { |element| element.is_a?(Relation) }
+
       records.flatten!(1)
       records.compact!
 
-      unless records.all?(klass)
+      unless records.all?(klass) && relations.all? { |relation| relation.klass == klass }
         raise ArgumentError, "You must only pass a single or collection of #{klass.name} objects to ##{__callee__}."
       end
 
-      spawn.excluding!(records)
+      spawn.excluding!(records + relations.flat_map(&:ids))
     end
     alias :without :excluding
```

It might also be good to add a test that asserts no extra queries when the relation is already loaded.

Reply:
@jonathanhefner thank you! I've pushed up these changes along with some test coverage.